jupyter-extension/jupyterlab_pachyderm/handlers.py

```diff
@@ -364,20 +389,52 @@ async def get(self):
-        self.settings["pfs_contents_manager"] = PFSManager(client=self.client)
-        self.settings["datum_contents_manager"] = DatumManager(client=self.client)
-        self.settings["pachyderm_pps_client"] = PPSClient(client=self.client)
 
     @tornado.web.authenticated
     async def put(self):
         try:
+            # Note: The auth workflow is for the backend to initiate the process by
+            # calling auth.get_oidc_login() which returns a url for the user to login
+            # with, and a state token that the backend can use to complete the workflow.
+            # Therefore, we send the url to the frontend for the user to login with and
+            # wait until the server has created a new session token, which we then retrieve.
             oidc_response = self.client.auth.get_oidc_login()
-            asyncio.create_task(self.get_token(oidc_response.state))
             response = oidc_response.to_json()
-            self.finish(response)
+            await self.finish(response)
+
+            # Unfortunately, there doesn't seem to be a way for us to use the synchronous
+            # version of client.auth.authenticate because it is blocking -- using it blocks
+            # the entire server until either the user successfully logs in or the OIDC login
+            # attempt times out. So, we need to manually do this Authenticate asynchronously.
+            async with grpc.aio.insecure_channel(self.client.address) as channel:
+                async_authenticate = channel.unary_unary(
+                    "/auth_v2.API/Authenticate",
+                    request_serializer=AuthenticateRequest.SerializeToString,
+                    response_deserializer=AuthenticateResponse.FromString,
+                )
+                try:
+                    response = await async_authenticate(
+                        AuthenticateRequest(oidc_state=oidc_response.state),
+                    )
+                except grpc.RpcError as err:
+                    # If the user opens several login attempts, we don't have a clear
+                    # mechanism to cull them after they have successfully logged in.
+                    # When the call inevitably times out, we don't raise and log a warning.
+                    details = err.details()
+                    if "OIDC state token expired" in details:
+                        get_logger().warning(details)
+                        return
+                    raise err
+                token = response.pach_token
+            self.client.auth_token = token
+
+            # Attempt to write new pachyderm context to config.
+            try:
+                write_config(self.client.address, self.client.root_certs, token)
+            except RuntimeError as e:
+                get_logger().error(f"Error updating local config: {e}.", exc_info=True)
+                raise tornado.web.HTTPError(500, f"Error updating local config: {e}.")
```

Comment:
i wonder if responding 500 might be misleading here. the client is successfully updated at this point but the config is not written to file. we could consider either reverting the client here, or maybe responding in a way that affirms the login attempt while warning that it's not persisted to the config file.